### PR TITLE
bst

### PR DESCRIPTION
**Pull request by agarwalvickey8**

### CODE REVIEW

File: binary_search_tree.rb [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+  attr_accessor :value, :left, :right
+  def initialize(value) 
+    @value = value
+    @left = nil
+    @right = nil
+  end
+end
+class BST
+  attr_accessor :root
+  def initialize()
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: no need of brackets

File: binary_search_tree.rb [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    @root=nil
+  end
+end
+def insert(value)
+   if @root == nil
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: if self.root.nil?

use self.root instead of @root

File: binary_search_tree.rb [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+     @root = Tree.new(value) 
+   else 
+     current_node = @root 
+     previous_node = @root
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: use self.root

File: binary_search_tree.rb [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    @root=nil
+  end
+end
+def insert(value)
+   if @root == nil
+     @root = Tree.new(value) 
+   else 
+     current_node = @root 
+     previous_node = @root 
+     while current_node != nil
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: while !current.node.nil?

File: binary_search_tree.rb [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+       previous_node.left = Tree.new(value)
+     else 
+       previous_node.right = Tree.new(value)
+     end
+   end
+end
+
+def largest_element(node=self.root)
+  if node == nil
+    return nil
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: return nil if node.nil?

File: binary_search_tree.rb [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+  else if node.right == nil
+    return node.value
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: return node.value if node.right.nil?

File: binary_search_tree.rb [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+  else
+    return largest_element(node.right)
+  end
+end
+
+def smallest_element(node=self.root)
+  if node == nil
+    return nil
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: return nil if node.nil?

File: binary_search_tree.rb [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+  else if node.left == nil
+    return node.value
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: return node.value if node.left.nil?

File: binary_search_tree.rb [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+  else
+    return smallest_element(root.left)
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: no need to write return at the last return
only smallest_element(root.left)

File: binary_search_tree.rb [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+def smallest_element(node=self.root)
+  if node == nil
+    return nil
+  else if node.left == nil
+    return node.value
+  else
+    return smallest_element(root.left) 
+def in_order(node = self.root)
+  if node == nil
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: return if node.nil?

File: binary_search_tree.rb [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+    return
+  end
+  in_order(node.left)
+  print(node.value)
+  print(" ")
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: print "#{node.value} "

File: binary_search_tree.rb [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+  in_order(node.right)
+end
+
+def pre_order(node=self.root)
+  if node == nil
+    return
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: return if node.nill?

File: binary_search_tree.rb [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+  print(" ")
+  in_order(node.right)
+end
+
+def pre_order(node=self.root)
+  if node == nil
+    return 
+  end
+    print(node.value)
+    print(" ")
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: print "#{node.value} "

File: binary_search_tree.rb [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+    pre_order(node.left)
+    pre_order(node.right)
+end
+
+def post_order(node=self.root)
+  if root == nil
+    return
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: return if node.nill?

File: binary_search_tree.rb [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+  end
+  post_order(node.left)
+  post_order(node.right)
+  print(node.value)
+  print(" ")
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: print "#{node.value} "

File: binary_search_tree.rb [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+end
+
+def in_order(node=self.root)
+  if node == nil
+    return
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: return if node.nil?

File: binary_search_tree.rb [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+  in_order(node.left)
+  print(node.value)
+  print(" ")
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: print "#{node.value} "

File: binary_search_tree.rb [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+  in_order(node.right)
+  end
+end  
+  
+def search_element(node = self.root,element)
+  if node == nil
+    return false
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: return false if node.nil?

File: binary_search_tree.rb [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+  end
+  if node.value == element
+    return true
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: return true if element == node.value

File: binary_search_tree.rb [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+  else if node.value > element
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: elsif element < node.value

File: binary_search_tree.rb [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+    search_element(node.left, element)
+  else
+    search_element(node.right, element)
+  end
+end
+
+def remove_element(value,node = self.root)
+  if node == nil
+    return nil
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: return nil if node.nil?

File: binary_search_tree.rb [hunk 22]
```diff
@@ -0,0 +1,191 @@
...
+  end
+  if node.value > value
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: if value < node.value

File: binary_search_tree.rb [hunk 23]
```diff
@@ -0,0 +1,191 @@
...
+    node.left = remove_element(value, node.left)
+  elsif node.value < value
+    node.right = remove_element(value, node.right)
+  else
+    if node.left != nil && node.right != nil
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: elsif node.left.nil? && node.right.nil?

File: binary_search_tree.rb [hunk 24]
```diff
@@ -0,0 +1,191 @@
...
+      temp = node
+      min_of_right_subtree = find_min(node.right)
+      node.value = min_of_right_subtree.value
+      node.right = remove_element(min_of_right_subtree.value, node.right)
+    elsif node.left != nil
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: elsif node.left.nil?

File: binary_search_tree.rb [hunk 25]
```diff
@@ -0,0 +1,191 @@
...
+      node = node.left
+    elsif node.right != nil
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: elsif node.right.nil?

File: binary_search_tree.rb [hunk 26]
```diff
@@ -0,0 +1,191 @@
...
+      node = node.right
+    else
+      node = nil
+    end
+  end
+  return node
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: node

~no return~

File: binary_search_tree.rb [hunk 27]
```diff
@@ -0,0 +1,191 @@
...
+puts "11.Quit"
+operation_num=gets.chomp()
+operation_num.to_i
+root=new.BST
+while(operation_num!=11)
+  case operation_num
+  when 1
+    elements=[]
+    print "How much values you want to add in BST:"
+    num_values=gets.chomp()
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: do as per defined that is comma-separated input

File: binary_search_tree.rb [hunk 28]
```diff
@@ -0,0 +1,191 @@
...
+    for i in 1..num_values
+      value=gets()
+      print(", ")
+      insert(value)
+  when 2
+    puts(largest_element(root))
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: largest_element function should return the element 
it should not print it

File: binary_search_tree.rb [hunk 29]
```diff
@@ -0,0 +1,191 @@
...
+  when 3
+    puts(smallest_element(root))
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: smallest_element function  should return the element
it should not print it

File: binary_search_tree.rb [hunk 30]
```diff
@@ -0,0 +1,191 @@
...
+  when 5
+    post_order(root)
+  when 6
+    level_order(root)
+  when 7
+    pre_order(root)
+  when 8
+    print("Enter element which you want to search:")
+    element=gets.chomp()
+    print(search_element(root,element)
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: search element should return the boolean type, it should not print it

File: binary_search_tree.rb [hunk 31]
```diff
@@ -0,0 +1,191 @@
...
+  when 9
+    print("Enter element which you want to remove:")
+    element=gets.chomp()
+    print(remove_element(root,element))
+  when 10
+    print(all_paths(root))
+  when 11
+    print("exit")
```
Comment (SparshAgrawal1) [inline binary_search_tree.rb]: when user decides to exit, you should also put the elements of the binary search tree in one separate file (output.txt)

**Review by SparshAgrawal1 — CHANGES_REQUESTED** [review]
Many changes are required

File: binary_search_tree.rb [hunk 32]
```diff
@@ -0,0 +1,189 @@
...
+    if name = element_file
+      load_bst()
+  when 12
+    elements_file(root)
+    print("exit")
+    loop_end = true
+  else 
+    "Default value"
+  end
+end
```
Comment (bv-hemant) [inline binary_search_tree.rb]: use tab not space

File: binary_search_tree.rb [hunk 33]
```diff
@@ -0,0 +1,189 @@
...
+      else 
+        current_node = current_node.right
+      end
+    end
+    if value < previous_node.value
+      previous_node.left = Tree.new(value)
+    else 
+      previous_node.right = Tree.new(value)
+    end
+  end
```
Comment (bv-hemant) [inline binary_search_tree.rb]: have you tested it?

**Review by bv-hemant — COMMENTED** [review]
test it. It will break

File: binary_search_tree.rb [hunk 34]
```diff
@@ -0,0 +1,190 @@
...
+	else 
+		current_node = self.root 
+		previous_node = self.root 
+		while !current_node.nil? 
+			previous_node = current_node 
+			if value < current_node.value
+				current_node = current_node.left
+			else
+				current_node = current_node.right
+			end
```
Comment (bv-hemant) [inline binary_search_tree.rb]: use ternary

File: binary_search_tree.rb [hunk 35]
```diff
@@ -0,0 +1,190 @@
...
+		else 
+			previous_node.right = Tree.new(value)
+		end
+	end
+end
+
+def largest_element(node=self.root)
+	return nil if node.nil?
+	return node.value if node.right.nil?
+	return largest_element(node.right)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: no need of return in last

File: binary_search_tree.rb [hunk 36]
```diff
@@ -0,0 +1,190 @@
...
+		end
+		if value < previous_node.value
+			previous_node.left = Tree.new(value)
+		else 
+			previous_node.right = Tree.new(value)
+		end
+	end
+end
+
+def largest_element(node=self.root)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: node = self.root

File: binary_search_tree.rb [hunk 37]
```diff
@@ -0,0 +1,190 @@
...
+	return nil if node.nil?
+	return node.value if node.right.nil?
+	return largest_element(node.right)
+end
+
+def smallest_element(node=self.root)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: node = slef.root

File: binary_search_tree.rb [hunk 38]
```diff
@@ -0,0 +1,190 @@
...
+	return nil if node.nil?
+	return node.value if node.left.nil?
+	smallest_element(root.left)
+end
+
+def in_order(node = self.root)
+	return if node.nil?
+	in_order(node.left)
+	print"#{node.value}"
```
Comment (bv-hemant) [inline binary_search_tree.rb]: print "#{node.value} "

File: binary_search_tree.rb [hunk 39]
```diff
@@ -0,0 +1,190 @@
...
+	print(" ")
```
Comment (bv-hemant) [inline binary_search_tree.rb]: no need

File: binary_search_tree.rb [hunk 40]
```diff
@@ -0,0 +1,190 @@
...
+	in_order(node.right)
+end
+
+def pre_order(node=self.root)
+	return if node.nil?
+	print"#{node.value}"
```
Comment (bv-hemant) [inline binary_search_tree.rb]: print "#{node.value} "

File: binary_search_tree.rb [hunk 41]
```diff
@@ -0,0 +1,190 @@
...
+
+def in_order(node = self.root)
+	return if node.nil?
+	in_order(node.left)
+	print"#{node.value}"
+	print(" ")
+	in_order(node.right)
+end
+
+def pre_order(node=self.root)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: node  = self.root

File: binary_search_tree.rb [hunk 42]
```diff
@@ -0,0 +1,190 @@
...
+	return if node.nil?
+	print"#{node.value}"
+  print(" ")
```
Comment (bv-hemant) [inline binary_search_tree.rb]: no need

**Review by bv-hemant — CHANGES_REQUESTED** [review]
test it properly

File: binary_search_tree.rb [hunk 43]
```diff
@@ -0,0 +1,190 @@
...
+
+class BST
+	attr_accessor :root
+
+  def initialize
+		@root = nil
+	end
+	def insert(value)
+		if self.root.nil?
+			self.root = Tree.new(value)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: TreeNode

File: binary_search_tree.rb [hunk 44]
```diff
@@ -0,0 +1,190 @@
+class Tree
```
Comment (bv-hemant) [inline binary_search_tree.rb]: TrreNode

File: binary_search_tree.rb [hunk 45]
```diff
@@ -0,0 +1,190 @@
...
+end
+
+class BST
+	attr_accessor :root
+
+  def initialize
+		@root = nil
+	end
+	def insert(value)
+		if self.root.nil?
```
Comment (bv-hemant) [inline binary_search_tree.rb]: indentation

File: binary_search_tree.rb [hunk 46]
```diff
@@ -0,0 +1,190 @@
...
+			else 
+				previous_node.right = Tree.new(value)
+			end
+		end
+	end
+
+	def largest_element(node = self.root)
+		return node if node.nil?
+		return node.value if node.right.nil?
+		return largest_element(node.right)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: no need of return

File: binary_search_tree.rb [hunk 47]
```diff
@@ -0,0 +1,190 @@
...
+		return node if node.nil?
+		return node.value if node.left.nil?
+		smallest_element(root.left)
+	end
+
+	def in_order(node = self.root)
+		return if node.nil?
+		in_order(node.left)
+		print "#{node.value}"
+		print(" ")
```
Comment (bv-hemant) [inline binary_search_tree.rb]: print "#{node.value} "

File: binary_search_tree.rb [hunk 48]
```diff
@@ -0,0 +1,190 @@
...
+		in_order(node.right)
+	end
+
+	def pre_order(node = self.root)
+		return if node.nil?
+		print "#{node.value}"
+		print(" ")
```
Comment (bv-hemant) [inline binary_search_tree.rb]: print "#{node.value} "

File: binary_search_tree.rb [hunk 49]
```diff
@@ -0,0 +1,190 @@
...
+		pre_order(node.left)
+		pre_order(node.right)
+	end
+
+	def post_order(node = self.root)
+		return if root.nil?
+		post_order(node.left)
+		post_order(node.right)
+		print "#{node.value}"
+		print(" ")
```
Comment (bv-hemant) [inline binary_search_tree.rb]: print "#{node.value} "

File: binary_search_tree.rb [hunk 50]
```diff
@@ -0,0 +1,190 @@
...
+	end  
+
+	def search_element(node = self.root,element)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: def search_element(node = self.root, element)

File: binary_search_tree.rb [hunk 51]
```diff
@@ -0,0 +1,190 @@
...
+		return false if node.nil?
+		return search_element(node.left, element) if element < node.value
+		return search_element(node.right, element) if node.value < element
+		true
+	end
+
+	def all_path(node = self.root, arr)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: what is arr? give proper naming

File: binary_search_tree.rb [hunk 52]
```diff
@@ -0,0 +1,190 @@
...
+		return if node.nil?
+		arr.push(node.value)
+		if(node.left.nil? && node.right.nil?)
+			print " " , arr
+		end
```
Comment (bv-hemant) [inline binary_search_tree.rb]: print " #{arr}" if node.left.nil? && node.right.nil?

File: binary_search_tree.rb [hunk 53]
```diff
@@ -0,0 +1,190 @@
...
+	end
+
+	def remove_element(value, node = self.root)
+		return nil if node.nil?
+		if value < node.value
+			node.left = remove_element(value, node.left)
+		elsif node.value < value
+			node.right = remove_element(value, node.right)
+		else
+			if(node.left.nil? && node.right.nil?)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: why brackets?

File: binary_search_tree.rb [hunk 54]
```diff
@@ -0,0 +1,190 @@
...
+			elsif node.right.nil?
+				node = node.right
+			else
+				node = nil
+			end
+		end
+		node
+	end
+
+	def elements_file(node = self.root)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: give proper naming. 
and fix this

File: binary_search_tree.rb [hunk 55]
```diff
@@ -0,0 +1,190 @@
...
+		elements_file(node.right)
+	end
+
+	def load_BST
+		root = BST.new
+		elements = fileobject.read()
+		for i in 1..elements.length
+			insert(value)
+		end
+	end
```
Comment (bv-hemant) [inline binary_search_tree.rb]: take input outside the class

File: binary_search_tree.rb [hunk 56]
```diff
@@ -0,0 +1,190 @@
...
+		return if node.nil?
+		elements_file(node.left)
+		fileobject.syswrite(node.value)
+		elements_file(node.right)
+	end
+
+	def load_BST
+		root = BST.new
+		elements = fileobject.read()
+		for i in 1..elements.length
```
Comment (bv-hemant) [inline binary_search_tree.rb]: elements.each { |element| insert(element) }

File: binary_search_tree.rb [hunk 57]
```diff
@@ -0,0 +1,190 @@
...
+			name = gets()
+			if name == element_file
+				load_bst
+			end
+		when 12
+			elements_file(root)
+			print("exit")
+			loop_end = true
+		else 
+			puts "Default value"
```
Comment (bv-hemant) [inline binary_search_tree.rb]: do this all things outside the class

File: binary_search_tree.rb [hunk 58]
```diff
@@ -0,0 +1,190 @@
...
+	puts "7.Print preorder of BST"
+	puts "8.Check whether element is present or not"
+	puts "9.Remove an element from BST"
+	puts "10.Print all the paths i.e starting from the root to the leaf"
+	puts "11.Enter the file name for previous values in BST"
+	puts "12.Quit"
+
+	root = BST.new
+	loop_end = false
+	while(!loop_end)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: while true

File: binary_search_tree.rb [hunk 59]
```diff
@@ -0,0 +1,190 @@
...
+		operation_num = gets.chomp()
```
Comment (bv-hemant) [inline binary_search_tree.rb]: why brackets

File: binary_search_tree.rb [hunk 60]
```diff
@@ -0,0 +1,190 @@
...
+		case operation_num.to_i
+		when 1
+			print("How much values you want to add in BST:")
+			num_values = gets.chomp
+			for i in 1..num_values.to_i
+				value = gets
+				print(", ")
+				insert(value.to_i)
+			end
```
Comment (bv-hemant) [inline binary_search_tree.rb]: do as per task

File: binary_search_tree.rb [hunk 61]
```diff
@@ -0,0 +1,190 @@
...
+		when 2
+			print(largest_element)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: if no largest element

File: binary_search_tree.rb [hunk 62]
```diff
@@ -0,0 +1,190 @@
...
+		when 3
+			print(smallest_element)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: if no smallest element

File: binary_search_tree.rb [hunk 63]
```diff
@@ -0,0 +1,190 @@
...
+		when 4
+			in_order(root)
+		when 5
+			post_order(root)
+		when 6
+			level_order(root)
+		when 7
+			pre_order(root)
+		when 8
```
Comment (bv-hemant) [inline binary_search_tree.rb]: why do you need to send root

File: binary_search_tree.rb [hunk 64]
```diff
@@ -0,0 +1,190 @@
...
+			print("Enter element which you want to search:")
+			element = gets.chomp()
```
Comment (bv-hemant) [inline binary_search_tree.rb]: why brackets

File: binary_search_tree.rb [hunk 65]
```diff
@@ -0,0 +1,190 @@
...
+			search_element(root, element.to_i)
+		when 9
+			print("Enter element which you want to remove:")
+			element = gets.chomp
+			print(remove_element(element.to_i))
+		when 10
+			arr = Array.new
```
Comment (bv-hemant) [inline binary_search_tree.rb]: proper naming and why you are sending root

File: binary_search_tree.rb [hunk 66]
```diff
@@ -0,0 +1,190 @@
...
+			all_paths(root, arr)
+		when 11
+			name = gets()
```
Comment (bv-hemant) [inline binary_search_tree.rb]: if file does not exist then

File: binary_search_tree.rb [hunk 67]
```diff
@@ -0,0 +1,190 @@
...
+			if name == element_file
+				load_bst
+			end
+		when 12
+			elements_file(root)
+			print("exit")
+			loop_end = true
+		else
```
Comment (bv-hemant) [inline binary_search_tree.rb]: rethink and fix this

File: binary_search_tree.rb (hunk 57 above)
Comment (bv-hemant) [inline binary_search_tree.rb]: what is this?

**Review by bv-hemant — CHANGES_REQUESTED** [review]

File: binary_search_tree.rb [hunk 68]
```diff
@@ -0,0 +1,180 @@
...
+	end
+end
+
+class BST
+	attr_accessor :root
+
+	def initialize
+		@root = nil
+	end
+	def insert(value)
```
Comment (bv-hemant) [inline binary_search_tree.rb:18]: add extra line between two function

File: binary_search_tree.rb [hunk 69]
```diff
@@ -0,0 +1,180 @@
...
+		if self.root.nil?
+			self.root = TreeNode.new(value) 
+		else 
+			current_node = self.root 
+			previous_node = self.root 
+			while current_node?
```
Comment (bv-hemant) [inline binary_search_tree.rb]: current_node

File: binary_search_tree.rb [hunk 70]
```diff
@@ -0,0 +1,180 @@
...
+	end  
+
+	def search_element(node = self.root, element)
+		return false if node.nil?
+		return search_element(node.left, element) if element < node.value
+		return search_element(node.right, element) if node.value < element
+		true
+	end
+
+	def all_path(node = self.root, elements_array)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: elements

File: binary_search_tree.rb [hunk 71]
```diff
@@ -0,0 +1,180 @@
...
+			elsif node.right.nil?
+				node = node.right
+			else
+				node = nil
+			end
+		end
+		node
+	end
+end
+fileobject = File.new("elements_file.txt", "a+")
```
Comment (bv-hemant) [inline binary_search_tree.rb]: file

File: binary_search_tree.rb [hunk 72]
```diff
@@ -0,0 +1,180 @@
...
+def add_elements_to_file(node = self.root)
+	return if node.nil?
+	add_elements_to_file(node.left)
+	fileobject.syswrite(node.value)
+	add_elements_to_file(node.right)
+end
+
+def load_BST
+	root = BST.new
+	elements = fileobject.read()
```
Comment (bv-hemant) [inline binary_search_tree.rb]: input and output files are same?

File: binary_search_tree.rb [hunk 73]
```diff
@@ -0,0 +1,180 @@
...
+puts "6.Print levelorder of BST"
+puts "7.Print preorder of BST"
+puts "8.Check whether element is present or not"
+puts "9.Remove an element from BST"
+puts "10.Print all the paths i.e starting from the root to the leaf"
+puts "11.Enter the file name for previous values in BST"
+puts "12.Quit"
+
+root = BST.new
+while(true)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: no need of bracket

File: binary_search_tree.rb [hunk 74]
```diff
@@ -0,0 +1,180 @@
...
+	operation_num = gets.chomp
+	case operation_num.to_i
+	when 1
+		puts "Enter values which you want to add in BST separated by comma:"
+		element_values = gets
+		element_values_array = element_values.split(",")
+		for element in element_values_array do
+			insert(element.to_i)
+		end
```
Comment (bv-hemant) [inline binary_search_tree.rb]: element_values_array.each { |element| bst.insert(element.to_i) }

**Review by bv-hemant — CHANGES_REQUESTED** [review]

File: binary_search_tree.rb [hunk 75]
```diff
@@ -0,0 +1,180 @@
...
+puts "5.Print postorder of BST"
+puts "6.Print levelorder of BST"
+puts "7.Print preorder of BST"
+puts "8.Check whether element is present or not"
+puts "9.Remove an element from BST"
+puts "10.Print all the paths i.e starting from the root to the leaf"
+puts "11.Enter the file name for previous values in BST"
+puts "12.Quit"
+
+root = BST.new
```
Comment (bv-hemant) [inline binary_search_tree.rb]: bst

File: binary_search_tree.rb [hunk 76]
```diff
@@ -0,0 +1,183 @@
...
+	end
+
+	def post_order(node = self.root)
+		return if root.nil?
+		post_order(node.left)
+		post_order(node.right)
+		print "#{node.value} "
+	end  
+
+	def search_element(node = self.root, element)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: def search_element( element, node = self.root)

File: binary_search_tree.rb [hunk 77]
```diff
@@ -0,0 +1,183 @@
...
+		return false if node.nil?
+		return search_element(node.left, element) if element < node.value
+		return search_element(node.right, element) if node.value < element
+		true
+	end
+
+	def all_path(node = self.root, elements)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: def all_path(elements, node = self.root)

File: binary_search_tree.rb [hunk 78]
```diff
@@ -0,0 +1,183 @@
...
+while true
+	operation_num = gets.chomp
+	case operation_num.to_i
+	when 1
+		puts "Enter values which you want to add in BST separated by comma:"
+		element_values = gets
+		element_values_array = element_values.split(",")
+		element_values_array.each { |element| bst.insert(element.to_i) }
+	when 2
+		bst ? print(bst.largest_element) : print("Currently BST is empty")
```
Comment (bv-hemant) [inline binary_search_tree.rb]: print bst ?  bst.largest_element : "Currently BST is empty"

File: tree.rb [hunk 79]
```diff
@@ -0,0 +1,189 @@
...
+		when 5
+			post_order(root)
+		when 6
+			level_order(root)
+		when 7
+			pre_order(root)
+		when 8
+			print("Enter element which you want to search:")
+			element = gets.chomp()
+			search_element(element)
```
Comment (bv-hemant) [inline tree.rb]: print

File: tree.rb [hunk 80]
```diff
@@ -0,0 +1,189 @@
...
+			element = gets.chomp
+			print(remove_element(element))
+		when 10
+			arr = Array.new
+			all_paths(root,arr)
+		when 11
+			name = gets()
+			if name = element_file
+				load_bst
+			end
```
Comment (bv-hemant) [inline tree.rb]: one liner

File: tree.rb [hunk 81]
```diff
@@ -0,0 +1,189 @@
...
+			print("Enter element which you want to search:")
+			element = gets.chomp()
+			search_element(element)
+		when 9
+			print("Enter element which you want to remove:")
+			element = gets.chomp
+			print(remove_element(element))
+		when 10
+			arr = Array.new
+			all_paths(root,arr)
```
Comment (bv-hemant) [inline tree.rb]: space after `,`

File: tree.rb [hunk 82]
```diff
@@ -0,0 +1,189 @@
...
+	puts "8.Check whether element is present or not"
+	puts "9.Remove an element from BST"
+	puts "10.Print all the paths i.e starting from the root to the leaf"
+	puts "11.Enter the file name for previous values in BST"
+	puts "12.Quit"
+
+	root = BST.new
+	loop_end = false
+	while(!loop_end)
+		operation_num = gets.chomp()
```
Comment (bv-hemant) [inline tree.rb]: why empty brackets?

File: tree.rb [hunk 83]
```diff
@@ -0,0 +1,189 @@
...
+			search_element(element)
+		when 9
+			print("Enter element which you want to remove:")
+			element = gets.chomp
+			print(remove_element(element))
+		when 10
+			arr = Array.new
+			all_paths(root,arr)
+		when 11
+			name = gets()
```
Comment (bv-hemant) [inline tree.rb]: why empty brackets?

**Review by bv-hemant — CHANGES_REQUESTED** [review]

File: binary_search_tree.rb [hunk 84]
```diff
@@ -0,0 +1,183 @@
...
+		true
+	end
+
+	def all_path(elements, node = self.root)
+		return if node.nil?
+		elements.push(node.value)
+		print " #{elements_array}" if node.left.nil? && node.right.nil?
+		all_path(node.left, elements)
+		all_path(node.right, elements)
+		arr.pop()
```
Comment (bv-hemant) [inline binary_search_tree.rb]: where is arr defined?

File: binary_search_tree.rb [hunk 85]
```diff
@@ -0,0 +1,183 @@
...
+		fileobject.write(" #{node.value}")
+		add_elements_to_file(node.right, fileobject)
+	end
+
+	def load_BST
+		file = File.open("file.txt","w+")
+		file_data = file.read
+		@elements = []
+		@elements = file_data.split("\n").map(&:to_i)
+		make_bst(self.elements)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: where is the function

File: binary_search_tree.rb [hunk 86]
```diff
@@ -0,0 +1,183 @@
...
+puts "6.Print levelorder of BST"
+puts "7.Print preorder of BST"
+puts "8.Check whether element is present or not"
+puts "9.Remove an element from BST"
+puts "10.Print all the paths i.e starting from the root to the leaf"
+puts "11.Enter the file name for previous values in BST"
+puts "12.Quit"
+
+bst = BST.new
+fileobject = File.open("file.txt", "w+")
```
Comment (bv-hemant) [inline binary_search_tree.rb]: why?

File: binary_search_tree.rb [hunk 87]
```diff
@@ -0,0 +1,183 @@
...
+	when 5
+		bst.post_order
+	when 6
+		bst.level_order
+	when 7
+		bst.pre_order
+	when 8
+		print("Enter element which you want to search:")
+		element = gets
+		bst.search_element(element.to_i)
```
Comment (bv-hemant) [inline binary_search_tree.rb]: print?

File: binary_search_tree.rb [hunk 88]
```diff
@@ -0,0 +1,183 @@
...
+	when 9
+		print("Enter element which you want to remove:")
+		element = gets.chomp
+		print(bst.remove_element(element.to_i))
```
Comment (bv-hemant) [inline binary_search_tree.rb]: no need of brackets in print

File: binary_search_tree.rb [hunk 89]
```diff
@@ -0,0 +1,183 @@
...
+	when 10
+		elements_array = Array.new
+		bst.all_paths(elements_array)
+	when 11
+		name = gets
+		if name == elements_file && elements_file?
+			bst.load_BST
+		end
```
Comment (bv-hemant) [inline binary_search_tree.rb]: one liner

File: binary_search_tree.rb [hunk 90]
```diff
@@ -0,0 +1,183 @@
...
+	when 12
+		bst.add_elements_to_file(fileobject)
```
Comment (bv-hemant) [inline binary_search_tree.rb:177]: change the file name

Comment (bv-hemant) [inline binary_search_tree.rb:177]: make the fileobject local

**Review by bv-hemant — CHANGES_REQUESTED** [review]

**Pull request closed without merging**